proposals/improved-interpolated-strings.md

```diff
@@ -0,0 +1,436 @@
+and more specialized scenarios such as logging frameworks, without incurring unnecessary allocations from formatting the string in the framework.
+
+## Motivation
+
+Today, string interpolation mainly lowers down to a call to `string.Format`. This, while general purpose, can be inefficient for a number of reasons:
+
+1. It boxes any struct arguments, unless the runtime has happened to introduce an overload of `string.Format` that takes exactly the correct types of arguments
+in exactly the correct order.
+    * This ordering is why the runtime is hesitant to introduce generic versions of the method, as it would lead to combinatoric explosion of generic instanciations
+    of a very common method.
+2. It has to allocate an array for the arguments in most cases.
+3. There is no opportunity to avoid instanciating the instance if it's not needed. Logging frameworks, for example, will recommend avoiding string interpolation
+because it will cause a string to be realized that may not be needed, depending on the current log-level of the application.
+4. It can never use `Span` or other ref struct types today, because ref structs are not allowed as generic type parameters, meaning that if a user wants to avoid
+copying to intermediate locations they have to manually format strings.
+
+Internally, the runtime has a type called `ValueStringBuilder` to help deal with the first 2 of these scenarios. They pass a stackalloc'd buffer to the builder,
+repeatedly call `AppendFormat` with every part, and then get a final string out. If the resulting string goes past the bounds of the stack buffer, they can then
+move to an array on the heap. However, this type is dangerous to expose directly, as incorrect usage could lead to a rented array to be double-disposed, which
+then will cause all sorts of undefined behavior in the program as two locations think they have sole access to the rented array. This proposal creates a way to
+use this type safely from native C# code by just writing an interpolated string literal, leaving written code unchanged while improving every interpolated string
+that a user writes. It also extends this pattern to allow for interpolated strings passed as arguments to other methods to use a builder pattern, defined by
+receiver of the method, that will allow things like logging frameworks to avoid allocating strings that will never be needed, and giving C# users familiar,
+convenient interpolation syntax.
+
+## Detailed Design
+
+### The builder pattern
+
+We introduce a new builder pattern that can represent an interpolated string passed as an argument to a method. The simple English of the pattern is as follows:
+
+When an _interpolated\_string\_expression_ is passed as an argument to a method, we look at the type of the parameter. If the parameter type has a static method
+`GetInterpolatedStringBuilder` that can invoked with 2 int parameters, `baseLength` and `formatHoleCount`, optionally takes a parameter the receiver is convertible to,
```

Comment:
Where will we handle all of the annoying questions here like:

- What accessibilities on the `static` method do we allow?
- Do the parameter names have to match?
- What about `static interface` members ... 

Do we want to do it here or later in a more detailed spec?